### DISCUSSION TITLE
SSL Server socket with micropythen pico

### DISCUSSION THREAD

**Discussion by sishari** [micropython/micropython#13534] [RP2040 / Pico]
Hello,
Can anybody help me!
I'm trying to create an SSL server socket with following:

context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
context.load_cert_chain("my.crt", "my.key")

After this line i get allways:
ValueError: invalid key

loop = asyncio.get_event_loop()
loop.create_task(asyncio.start_server(serve, "0.0.0.0", 80,ssl = context))

I dont know what is the problem...

**Comment by sophiedegran** [comment]
Hi
In micropython certificate have to be in DER format not PEM
have a look to microdot tls example
Hope that help

**Reply by sishari** [reply]
Hello,

I have try the example but i get agian invalid key.
That's how I tried it:
1.
openssl req -x509 -newkey rsa:4096 -nodes -out cert.pem -keyout key.pem -days 365
openssl x509 -in cert.pem -out cert.der -outform DER
openssl rsa -in key.pem -out key.der -outform DER
2.
context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
context.load_cert_chain("cert.der", "key.der")

I tried it on a rpi pico with thonny

What is the problem????

**Reply by sishari** [reply]
I read that in the doku:
cadata is a bytes object containing the CA certificate chain (in DER format) that will validate the peer’s certificate. Currently only a single DER-encoded certificate is supported.

How do i create a single certificate ???

Thanks !!!

**Reply by sophiedegran** [reply]
It should work ... I suggest you to try with microdot directly  (many people used it and it works)
After you will have specifiq infos on the old forum (forum.micropython.org)
Tell me..

**Reply by mendenm** [reply]
Note that the documentation says the data provided isn't the file name, it is the bytes object with the cert.  Have you tried `context.load_cert_chain(open("cert.der", "rb").read(), open("key.der", "rb").read())`?

**Reply by sishari** [reply]
context.load_cert_chain(open("cert.der", "rb").read(), open("key.der", "rb").read())
Is the same problem... invalid key

I think my cert files are ok. Maybe someone can check my cert files to see if it is correct. (I have changed from *.der to *.txt)
[key.txt](https://github.com/micropython/micropython/files/14076129/key.txt)
[cert.txt](https://github.com/micropython/micropython/files/14076130/cert.txt)

**Reply by sishari** [reply]
Now i have tried microdot but with exact the same problem.
invalid key

**Reply by sophiedegran** [reply]
Ok so it's a  problem of generation of cert..

**Comment by sishari** [comment]
> It should work ... I suggest you to try with microdot directly (many people used it and it works) After you will have specifiq infos on the old forum (forum.micropython.org) Tell me..
 I dont know? In the microdot.py is the same code as mine.

**Reply by sophiedegran** [reply]
Try your cert under microdot ..
I suppose you're under micropython 1.22 STABLE (mini), so install just the tls example in microdot 2.0.3 ... and they dependances

**Reply by sishari** [reply]
For your info.
I have the v1.22 but it was the same problem.
I think the problem is the key but I have no idea what exactly is the problem.
I hope Carglglz can explain this.....

**Comment by Carglglz** [comment]
> 1. openssl req -x509 -newkey rsa:4096 -nodes -out cert.pem -keyout key.pem -days 365
openssl x509 -in cert.pem -out cert.der -outform DER
openssl rsa -in key.pem -out key.der -outform DER


@sishari  try this instead 

```
$ openssl pkey -in rsa_key.pem -out rsa_key.der -outform DER
$ openssl x509 -in rsa_cert.pem -out rsa_cert.der -outform DER
```
or even better 

Use elliptic curve key/cert pairs:
```
$ openssl ecparam -name prime256v1 -genkey -noout -out ec_key.der -outform DER
$ openssl req -new -x509 -key ec_key.der -out ec_cert.der -outform DER -days 365 -nodes 
```

**Reply by sishari** [reply]
Hello,

You are the best!
With the ec_key it works now!!!
Can you explain to me what is different now?

**Reply by sophiedegran** [reply]
Nice!

**Reply by Carglglz** [reply]
@sishari I don't know for sure but I think it is related to this https://en.wikipedia.org/wiki/PKCS i.e. `openssl rsa` using an old standard and `mbedtls` (MicroPython SSL backend) not accepting it 🤔, see https://github.com/micropython/micropython/pull/8988#issuecomment-1896230864
